### PR TITLE
[fix bug2815]check the network before cold migrate vm

### PR DESCRIPTION
**Pull request by HeathHose**

### CODE REVIEW

**Comment by HeathHose** [comment]
@ZStack-Robot 
The two clusters must access each other in l2 network before cold migrate vm
https://github.com/zstackio/issues/issues/2815

**Comment by HeathHose** [comment]
## Analysis
Check l2 network in clusters before cold migrate the vm. When handle APILocalStorageGetVolumeMigratableHostsMsg and APILocalStorageMigrateVolumeMsg,add the check.

File: plugin/localstorage/src/main/java/org/zstack/storage/primary/local/LocalStorageBase.java [hunk 1]
```diff
@@ -195,7 +197,55 @@ private void handle(APILocalStorageGetVolumeMigratableHostsMsg msg) {
         TypedQuery<HostVO> hq = dbf.getEntityManager().createQuery(sql, HostVO.class);
         hq.setParameter("uuids", hostUuids);
         hq.setParameter("hstatus", HostStatus.Connected);
-        List<HostVO> hosts = hq.getResultList();
+        LinkedList<HostVO> hosts = new LinkedList<>(hq.getResultList());
+
+        //When migrate the rootVolume, need to check if the network environment meets the requirement of vm running after migrate
+        new SQLBatch() {
```
Comment (zxwing) [inline plugin/localstorage/src/main/java/org/zstack/storage/primary/local/LocalStorageBase.java]: 既然用了SQLBatch, 把前面所有逻辑都放到SQLBatch里面

File: plugin/localstorage/src/main/java/org/zstack/storage/primary/local/LocalStorageApiInterceptor.java [hunk 2]
```diff
@@ -103,25 +112,57 @@ private void validate(APILocalStorageMigrateVolumeMsg msg) {
...
-
-            SimpleQuery<VolumeVO> vq = dbf.createQuery(VolumeVO.class);
-            vq.add(VolumeVO_.type, Op.EQ, VolumeType.Data);
-            vq.add(VolumeVO_.vmInstanceUuid, Op.EQ, vol.getVmInstanceUuid());
-            long count = vq.count();
-            if (count != 0) {
-                throw new ApiMessageInterceptionException(operr("the volume[uuid:%s] is the root volume of the vm[uuid:%s]. Currently the vm still" +
-                        " has %s data volumes attached, please detach them before migration", vol.getUuid(), vol.getVmInstanceUuid(), count));
-            }
+            new SQLBatch() {
```
Comment (zxwing) [inline plugin/localstorage/src/main/java/org/zstack/storage/primary/local/LocalStorageApiInterceptor.java]: 同样的，所有逻辑都放到SQLBatch里面

Comment (zxwing) [inline plugin/localstorage/src/main/java/org/zstack/storage/primary/local/LocalStorageApiInterceptor.java]: 另外check逻辑太长了，要加一些注释说明

**Pull request merged**